### PR TITLE
P4-2067 Add NOMIS Mappings and Importer

### PR DESCRIPTION
**Pull request by cwrw**
### Jira link

[P4-2067](https://dsdmoj.atlassian.net/browse/P4-2067)

### What?
- [x] Add Framework NOMIS mapping table and model
- [x] Add a mapping importer for each type of NOMIS mapping
- [x] Add NOMIS Mapping Importer

### Why?

NOMIS alerts, personal care needs as well as reasonable adjustments for a person will help users fill out the Person escort record more accurately. Each question in the PER is associated to the relevant mappings of NOMIS resource codes, or set as a fallback question if a person has NOMIS resources that are not associated to any question.

To allow users to view those associations, add an Importer which pulls in all the resources for a person, and maps them appropriately to the NOMIS response.

NOMIS codes are not unique to a question, so we need to make sure we associate mappings to each and every question a NOMIS code appears in. If this NOMIS code is not in out framework, we will group it under the fallback response with the correct type.

Failures of importing and validation/persisting of NOMIS mappings will be addressed in further tickets.
### Have you? (optional)

- Updated API docs if necessary: will happen in subsequent PRs

### Deployment risks (optional)

- The latest framework has not been added to production ENVs so we will not be using this importer yet

### CODE REVIEW

File: app/services/framework_nomis_mappings/importer.rb [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    def grouped_framework_nomis_codes
+      framework_nomis_codes.group_by(&:code_type)
+    end
+
+    def fallback_nomis_codes
+      @fallback_nomis_codes ||= framework_nomis_codes.select(&:fallback?)
+    end
+
+    def nomis_code_ids_to_mappings
+      @nomis_code_ids_to_mappings ||= begin
```
Comment (cwrw) [inline app/services/framework_nomis_mappings/importer.rb:61]: add test and account for duplicate NOMIS alerts

File: app/services/framework_nomis_mappings/importer.rb [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+          nomis_code_ids = responses_to_codes[response.id]&.pluck(:nomis_code_id)
+          response.framework_nomis_mappings = nomis_code_ids_to_mappings.slice(*nomis_code_ids).values.flatten
+        end
+      end
+    end
+
+  private
+
+    def persist_framework_nomis_mappings
+      @persist_framework_nomis_mappings ||= begin
```
Comment (smoothcontract) [inline app/services/framework_nomis_mappings/importer.rb:28]: I can't see this being called more than once, so no need to memoize?

Comment (smoothcontract) [reply app/services/framework_nomis_mappings/importer.rb:28]: Ignore that comment, I found another reference

File: app/services/framework_nomis_mappings/importer.rb [hunk 3]
```diff
@@ -0,0 +1,81 @@
+# frozen_string_literal: true
+
+module FrameworkNomisMappings
+  class Importer
+    attr_reader :person, :framework_responses, :framework_nomis_codes
```
Comment (smoothcontract) [inline app/services/framework_nomis_mappings/importer.rb:5]: Blank line here pls

File: app/services/framework_nomis_mappings/importer.rb [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+            end
+          elsif mapping_nomis_fallback
+            hash[mapping_nomis_fallback.id] = hash[mapping_nomis_fallback.id].to_a + [mapping]
+          end
+        end
+      end
+    end
+
+    def responses_to_codes
+      @responses_to_codes ||= framework_responses.joins(:framework_nomis_codes).select('framework_responses.id as id, framework_nomis_codes.id as nomis_code_id').group_by(&:id)
```
Comment (smoothcontract) [inline app/services/framework_nomis_mappings/importer.rb:78]: Not sure why this needs a `group_by` clause?

File: db/schema.rb [hunk 5]
```diff
@@ -162,6 +162,27 @@
...
 
+  create_table "framework_nomis_mappings", id: :uuid, default: -> { "gen_random_uuid()" }, force: :cascade do |t|
+    t.jsonb "raw_nomis_mapping", null: false
+    t.string "code", null: false
+    t.string "code_type", null: false
+    t.text "code_description"
+    t.text "comments"
+    t.date "start_date"
+    t.date "end_date"
+    t.date "creation_date"
```
Comment (smoothcontract) [inline db/schema.rb:173]: Is `creation_date` different from `created_at` ?

File: spec/services/framework_nomis_mappings/importer_spec.rb [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+    allow(NomisClient::Alerts).to receive(:get).and_return([nomis_alert])
+    allow(NomisClient::PersonalCareNeeds).to receive(:get).and_return(nomis_personal_care_needs)
+    allow(NomisClient::ReasonableAdjustments).to receive(:get).and_return(nomis_reasonable_adjustments)
+  end
+
+  it 'persists alerts, reasonable adjustments and personal care needs NOMIS mappings' do
+    framework_responses = FrameworkResponse.where(id: [framework_response1.id, framework_response2.id])
+    framework_nomis_codes = framework_responses.flat_map(&:framework_nomis_codes)
+
+    expect { described_class.new(person: person, framework_responses: framework_responses, framework_nomis_codes: framework_nomis_codes).call }.to change(FrameworkNomisMapping, :count).by(5)
```
Comment (smoothcontract) [inline spec/services/framework_nomis_mappings/importer_spec.rb:90]: Could you add a comment to explain why this adds 5 mappings please?

File: spec/services/framework_nomis_mappings/importer_spec.rb [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+    allow(NomisClient::Alerts).to receive(:get).and_return([])
+    allow(NomisClient::PersonalCareNeeds).to receive(:get).and_return([])
+    allow(NomisClient::ReasonableAdjustments).to receive(:get).and_return([])
+    framework_responses = FrameworkResponse.where(id: [framework_response1.id, framework_response2.id])
+    framework_nomis_codes = framework_responses.flat_map(&:framework_nomis_codes)
+
+    expect { described_class.new(person: person, framework_responses: framework_responses, framework_nomis_codes: framework_nomis_codes).call }.not_to change(FrameworkNomisMapping, :count)
+  end
+
+  it 'does nothing is no person present' do
```
Comment (smoothcontract) [inline spec/services/framework_nomis_mappings/importer_spec.rb:157]: `is` => `if`

File: spec/services/framework_nomis_mappings/importer_spec.rb [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+  end
+
+  it 'does nothing if no framework responses present' do
+    framework_responses = FrameworkResponse.where(id: [framework_response1.id, framework_response2.id])
+    framework_nomis_codes = framework_responses.flat_map(&:framework_nomis_codes)
+
+    expect { described_class.new(person: person, framework_responses: [], framework_nomis_codes: framework_nomis_codes).call }.not_to change(FrameworkNomisMapping, :count)
+  end
+
+  it 'does nothing is no framework NOMIS codes present' do
```
Comment (smoothcontract) [inline spec/services/framework_nomis_mappings/importer_spec.rb:171]: `is` => `if`

File: spec/services/framework_nomis_mappings/personal_care_needs_spec.rb [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+  end
+
+  it 'ignores personal care needs with status "I"' do
+    allow(NomisClient::PersonalCareNeeds).to receive(:get).and_return([nomis_personal_care_need(problem_status: 'I')])
+    mappings = described_class.new(prison_number: 'A9127EK').call
+
+    expect(mappings).to be_empty
+  end
+
+  it 'ignores personal care needs with status "EBS"' do
```
Comment (smoothcontract) [inline spec/services/framework_nomis_mappings/personal_care_needs_spec.rb:48]: Might be nice to have a comment somewhere indicating what `I` and `EBS` magic values represent

**Review by smoothcontract — APPROVED** [review]
A lot going on here so is a big PR. I think I follow it though and have left a few trivial comments for consideration.

File: app/services/framework_nomis_mappings/importer.rb (hunk 4 above)
Comment (cwrw) [reply app/services/framework_nomis_mappings/importer.rb:78]: We need to group the NOMIS codes by the response to make it easier and more efficient to lookup the NOMIS codes later on by response id, where now its through a hash key rather than having to select those every time. I added this to avoid doing more than one query to get those values, and instead have this in memory.

File: db/schema.rb (hunk 5 above)
Comment (cwrw) [reply db/schema.rb:173]: we wanted to differentiate the `created_at` coming from NOMIS for some resources from the `created_at` datetime table value to avoid confusion, so we went with `creation_date` instead

Comment (cwrw) [reply db/schema.rb:173]: It also ties nicely with `expiry_date`

File: spec/services/framework_nomis_mappings/importer_spec.rb (hunk 6 above)
Comment (cwrw) [reply spec/services/framework_nomis_mappings/importer_spec.rb]: Not sure I understand, this test checks that all the resources defined earlier are persisted, could you elaborate on where the uncertainty is and will add a comment

File: app/services/framework_nomis_mappings/importer.rb (hunk 4 above)
Comment (smoothcontract) [reply app/services/framework_nomis_mappings/importer.rb:79]: Awesome, thanks for the explanation 👍

File: db/schema.rb (hunk 5 above)
Comment (smoothcontract) [reply db/schema.rb:173]: Fair enough 👍

File: spec/services/framework_nomis_mappings/importer_spec.rb [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+    allow(NomisClient::Alerts).to receive(:get).and_return(nomis_alerts)
+    allow(NomisClient::PersonalCareNeeds).to receive(:get).and_return(nomis_personal_care_needs)
+    allow(NomisClient::ReasonableAdjustments).to receive(:get).and_return(nomis_reasonable_adjustments)
+  end
+
+  it 'persists alerts, reasonable adjustments and personal care needs NOMIS mappings' do
+    framework_responses = FrameworkResponse.where(id: [framework_response1.id, framework_response2.id])
+    framework_nomis_codes = framework_responses.flat_map(&:framework_nomis_codes)
+
+    expect { described_class.new(person: person, framework_responses: framework_responses, framework_nomis_codes: framework_nomis_codes).call }.to change(FrameworkNomisMapping, :count).by(6)
```
Comment (smoothcontract) [inline spec/services/framework_nomis_mappings/importer_spec.rb:103]: My question before was around why this expects 6 mappings to be created? I'm guessing, perhaps it's because there are two nomis alerts, two reasonable adjustments and two personal care needs? Or maybe something else.. but a little explanation here would be helpful for future readers.

File: spec/services/framework_nomis_mappings/personal_care_needs_spec.rb (hunk 9 above)
Comment (smoothcontract) [reply spec/services/framework_nomis_mappings/personal_care_needs_spec.rb]: 👍

File: spec/services/framework_nomis_mappings/importer_spec.rb (hunk 10 above)
Comment (smoothcontract) [reply spec/services/framework_nomis_mappings/importer_spec.rb:103]: Updated spec description makes sense, thanks for addressing 👍

**Review by smoothcontract — APPROVED** [review]
LGTM 👍
